pwiz/data/vendor_readers/Waters/SpectrumList_Waters.cpp

```diff
@@ -467,10 +471,18 @@ PWIZ_API_DECL void SpectrumList_Waters::getCombinedSpectrumData(int function, in
     auto mzItr = &mz[0], intensityItr = &intensity[0], driftTimeItr = &driftTime[0];
     for (int scan = 0; scan < numScansInBlock; ++scan)
     {
-        double dt = rawdata_->GetDriftTime(function, scan);
+        double dt;
+        if (config_.reportSonarBins)

```

Comment:
I'm not sure about this new reader config parameter. Shouldn't we just change the SONAR representation to the new recommend mzML representation (2 arrays for lower and upper quad m/z instead of 1 array for drift time)? Then you shouldn't need to make thousands of CLI calls to sonarBinToMzRange, right?

Reply:
From Skyline's point of view, it's much nicer to say "I've determined and cached the integer bin range of interest for each precursor, so just give me those bin numbers rather than making me deal with two values per bin."

Also those upper/lower quad m/z values are necessarily going to be fictions - there's no hard cutoff. The current implementation seems more, what, truthful.

Reply:
So you're making one sonarBinToMzRange call per precursor? How long does it take for 100k precursors?

Reply:
That's right, and using that bin number with Skyline's existing "ion mobility" filtering capability for chromatogram extraction, where bin number is the filtered value rather than ion mobility. Minimal code changes required on the Skyline side.

Hadn't thought to time that call specifically - why do you ask?

Reply:
Just seems like a lot of ferrying back and forth through the CLI layer. But if it's not a noticeably long time then I expect you're right and using bin number directly makes sense.

Reply:
The alternative is allocating, copying, and passing 1/3 again as much data per each of bazillions of measurements (mz,intensity,quadLow,quadHigh vs mz/intensity/bin) and checking twice as many values in the yet to be written filtering code. That just doesn't feel like a win, But I can't back that up with data, no.

Reply:
Yeah, but it turns out we've never actually needed to handle the 3 array IMS representation in mzML so I'm not too worried about 4-array SONAR in mzML.

Reply:
That's sound logic for 3 array IMS: I don't see that anybody's complained about it and it's been around for some time. But given Hans' long-standing interest in proper SONAR representation in mzML, should we run this decision by Hans? If he says that eventually he expects Skyline to support 4-array SONAR mzML, that may have implications for design choices now (e.g. do we want Skyline to have 2 filtering schemes, or do we want to have some kind of translation function in SpectrumList_IonMobility so Skyline can always treat it as bin numbers).

Reply:
If it really became an issue I'd probably just write some code to inspect the file and reverse engineer the bin setup. Sure, it's a performance hit, but mzML vs raw is already a performance hit so I'm not too bothered. But I think the dynamic is going to be the same with these large files as with other large files - folks will generally just avoid conversion to mzML when Skyline can read the original.